File: app/src/main/java/social/entourage/android/user/edit/UserEditPresenter.java
package social.entourage.android.user.edit;

import android.support.v4.util.ArrayMap;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import social.entourage.android.api.UserRequest;
import social.entourage.android.api.UserResponse;
import social.entourage.android.api.model.User;
import social.entourage.android.authentication.AuthenticationController;

/**
 * Created by mihaiionescu on 01/11/16.
 */


public class UserEditPresenter {

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    private final UserEditFragment fragment;
    private final UserRequest userRequest;
    private final AuthenticationController authenticationController;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    @Inject
    public UserEditPresenter(final UserEditFragment fragment,
                         final UserRequest userRequest,
                         final AuthenticationController authenticationController) {
        this.fragment = fragment;
        this.userRequest = userRequest;
        this.authenticationController = authenticationController;
    }

    // ----------------------------------
    // PUBLIC METHODS
    // ----------------------------------

    public void updateUser(final User user) {
        if (fragment != null) {
            Call<UserResponse> call = userRequest.updateUser(user.getArrayMapForUpdate());
            call.enqueue(new Callback<UserResponse>() {
                @Override
                public void onResponse(Call<UserResponse> call, Response<UserResponse> response) {
                    if (response.isSuccessful()) {
                        //update the logged user
                        authenticationController.saveUser(response.body().getUser());
                        authenticationController.saveUserPhoneAndCode(user.getPhone(), user.getSmsCode());
                        //inform the fragment
                        fragment.onUserUpdated(response.body().getUser());
                    }
                    else {
                        fragment.onUserUpdated(null);
                    }
                }

                @Override
                public void onFailure(Call<UserResponse> call, Throwable t) {
                    fragment.onUserUpdated(null);
                }
            });
        }
    }

    public void saveNewPassword(final String newPassword) {
        if (fragment != null) {
            ArrayMap<String, Object> userMap = new ArrayMap<>();
            userMap.put("sms_code", newPassword);
            Call<UserResponse> call = userRequest.updateUser(userMap);
            call.enqueue(new Callback<UserResponse>() {
                @Override
                public void onResponse(Call<UserResponse> call, Response<UserResponse> response) {
                    if (response.isSuccessful()) {
                        //inform the fragment
                        User user = authenticationController.getUser();
                        if (user != null) {
                            authenticationController.saveUserPhoneAndCode(user.getPhone(), newPassword);
                        }
                        fragment.onSaveNewPassword(newPassword);
                    }
                    else {
                        fragment.onSaveNewPassword(null);
                    }
                }

                @Override
                public void onFailure(Call<UserResponse> call, Throwable t) {
                    fragment.onSaveNewPassword(null);
                }
            });
        }
    }

    public void deleteAccount() {
        if (fragment != null) {
            Call<UserResponse> call = userRequest.deleteUser();
            call.enqueue(new Callback<UserResponse>() {
                @Override
                public void onResponse(final Call<UserResponse> call, final Response<UserResponse> response) {
                    if (response.isSuccessful()) {
                        fragment.onDeletedAccount(true);
                    }
                    else {
                        fragment.onDeletedAccount(false);
                    }
                }

                @Override
                public void onFailure(final Call<UserResponse> call, final Throwable t) {
                    fragment.onDeletedAccount(false);
                }
            });
        }
    }

}

File: app/src/main/java/social/entourage/android/user/edit/UserEditPresenter.kt
package social.entourage.android.user.edit

import android.os.Build
import androidx.collection.ArrayMap
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import social.entourage.android.EntourageApplication
import social.entourage.android.api.request.UserRequest
import social.entourage.android.api.request.UserResponse
import social.entourage.android.api.model.User
import social.entourage.android.authentication.AuthenticationController
import social.entourage.android.tools.log.AnalyticsEvents
import social.entourage.android.user.UserFragment
import social.entourage.android.user.edit.place.UserEditActionZoneFragment
import social.entourage.android.user.edit.place.UserEditActionZoneFragmentCompat
import javax.inject.Inject

/**
 * Created by mihaiionescu on 01/11/16.
 */
class UserEditPresenter @Inject constructor(
        private val fragment: UserEditFragment?,
        private val userRequest: UserRequest,
        private val authenticationController: AuthenticationController) {

    var editedUser: User? = null
        get() = field ?: authenticationController.me?.clone().also { field = it}
        private set

    // ----------------------------------
    // PUBLIC METHODS
    // ----------------------------------
    fun updateUser(user: User) {
        userRequest.updateUser(user.arrayMapForUpdate).enqueue(object : Callback<UserResponse> {
            override fun onResponse(call: Call<UserResponse>, response: Response<UserResponse>) {
                if (response.isSuccessful) {
                    response.body()?.user?.let {
                        //update the logged user
                        authenticationController.saveUser(it)
                        authenticationController.saveUserPhoneAndCode(user.phone, user.smsCode)
                        //inform the fragment
                        fragment?.onUserUpdated(it)
                    }
                } else {
                    fragment?.onUserUpdateError()
                }
            }

            override fun onFailure(call: Call<UserResponse>, t: Throwable) {
                fragment?.onUserUpdateError()
            }
        })
    }

    fun saveNewPassword(newPassword: String) {
        val userMap = ArrayMap<String, Any>()
        userMap["sms_code"] = newPassword
        userRequest.updateUser(userMap).enqueue(object : Callback<UserResponse> {
            override fun onResponse(call: Call<UserResponse>, response: Response<UserResponse>) {
                if (response.isSuccessful) {
                    //inform the fragment
                    authenticationController.me?.phone?.let { phone ->
                        authenticationController.saveUserPhoneAndCode(phone, newPassword)
                    }
                    editedUser?.smsCode = newPassword
                    fragment?.onSaveNewPassword()
                } else {
                    fragment?.onSavePasswordError()
                }
            }

            override fun onFailure(call: Call<UserResponse>, t: Throwable) {
                fragment?.onSavePasswordError()
            }
        })
    }

    fun deleteAccount() {
        userRequest.deleteUser().enqueue(object : Callback<UserResponse> {
            override fun onResponse(call: Call<UserResponse>, response: Response<UserResponse>) {
                fragment?.onDeletedAccount(response.isSuccessful)
            }

            override fun onFailure(call: Call<UserResponse>, t: Throwable) {
                fragment?.onDeletedAccount(false)
            }
        })
    }

    fun storeActionZone(ignoreActionZone: Boolean) {
        authenticationController.isIgnoringActionZone = ignoreActionZone

        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
            (fragment?.parentFragmentManager?.findFragmentByTag(UserEditActionZoneFragmentCompat.TAG) as UserEditActionZoneFragmentCompat?)?.let { userEditActionZoneFragmentCompat ->
                if (!userEditActionZoneFragmentCompat.isStateSaved) {
                    userEditActionZoneFragmentCompat.dismiss()
                }
            }
        } else {
            (fragment?.parentFragmentManager?.findFragmentByTag(UserEditActionZoneFragment.TAG) as UserEditActionZoneFragment?)?.let { userEditActionZoneFragment ->
                if (!userEditActionZoneFragment.isStateSaved) {
                    userEditActionZoneFragment.dismiss()
                }
            }
        }
    }

    fun deleteSecondaryAddress() {
        userRequest.deleteSecondaryAddressLocation()
                .enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    authenticationController.me?.let { me->
                        me.addressSecondary = null
                        authenticationController.saveUser(me)
                        fragment?.initUserData()
                    }
                }
            }
            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {}
        })
    }

    fun updateUser() {
        AnalyticsEvents.logEvent(AnalyticsEvents.EVENT_USER_SAVE)
        editedUser?.let { user->
            // If we have an user fragment in the stack, let it handle the update
            (fragment?.parentFragmentManager?.findFragmentByTag(UserFragment.TAG) as UserFragment?)?.saveAccount(user)
                    ?: run {
                        updateUser(user)
                    }
        }
    }

    fun initEditedUser() {
        authenticationController.me?.let { me ->
            editedUser?.let { user ->
                user.avatarURL = me.avatarURL
                user.partner = me.partner
                user.address = me.address
                user.addressSecondary = me.addressSecondary
                user.interests.clear()
                user.interests.addAll(me.interests)
                user.goal = me.goal
                fragment?.initUserData()
            }
        }
    }
}